start of web tests for the cdn/dev

smalls opened the pull request:
Still to do is to have these automatically run as part of either the current travis situation or sauce.

To run these on your local machine:

> cd arcs-cdn/dev
dev> npm install
dev> ../node_modules/selenium-standalone/bin/selenium-standalone install
dev> ../node_modules/selenium-standalone/bin/selenium-standalone start

in new shell (since selenium-standalone start will block):
> cd arcs-cdn/dev
dev> ../node_modules/webdriverio/bin/wdio test/wdio.conf.js

@shaper I was going to ask you to review this if you have time.

smalls committed:
very basic selenium test
To set up:
> npm install webdriverio chromedriver
> curl -O http://selenium-release.storage.googleapis.com/3.5/selenium-server-standalone-3.5.3.jar
> java -jar -Dwebdriver.chrome.driver=../../node_modules/chromedriver/bin/chromedriver ~/Downloads/selenium-server-standalone-3.5.3.jar
> node starter-test.js
(the last two commands will need to be run in separate shells)

smalls committed:
start on using wdio as the test runner

install selenium-standalone into node (no longer do you need to install it)
dev> npm install selenium-standalone
dev> ../node_modules/selenium-standalone/bin/selenium-standalone install
dev> ../node_modules/selenium-standalone/bin/selenium-standalone start

to run tests (although there are none yet)
dev> ../node_modules/.bin/wdio wdio.conf.js

smalls committed:
very simple test running with headless chrome
> ../node_modules/webdriverio/bin/wdio wdio.conf.js

smalls committed:
small fixes for the webdriver test
- consolidate test logic in the test spec (deleting the old standalone test)
- organize webdriver dependencies into dev
- ignore errorShots

smalls committed:
a utility method to descend across shadow roots

smalls committed:
successfully waiting for the dots to stop dancing

smalls committed:
in-progress on retrieving a suggestion via text
This path runs into issues with async execution; it may be time to back out
and try another fork.

smalls committed:
remove async/await from test method

smalls committed:
accepting suggestions works, and is now factored into a method

smalls committed:
the test now gets to the restaurant detail page
- pierceShadows will now return all matching results (a convenience method,
  pierceShadowsSingle asserts that there's only 1 result)
- continue factoring heavy lifting out of the test and into convenience
  methods (e.g, for accepting suggestions, or clicking in a specific
  slot/particle).

smalls committed:
clean up comments & function names

smalls committed:
drive the test back to the point of selecting a time at which the user is free
Handle some errors a bit more gracefully (by logging and retrying).

smalls committed:
clarify timeouts for debugging

smalls committed:
consolidate test resources in the test directory
Since wdio.conf.js has moved, now to run tests use:
dev> ../node_modules/webdriverio/bin/wdio test/wdio.conf.js

shaper commented on dev/test/selenium-utils.js:
```diff
@@ -0,0 +1,82 @@
+/** As #pierceShadows, but asserts that only a single result is found. */
+function pierceShadowsSingle(querySelectors) {
+  let result = _pierceShadows(document, querySelectors);
+  if (result.length>1) {
```
Run `prettier` with `--single-quote` on this?

shaper commented on dev/test/selenium-utils.js:
```diff
@@ -0,0 +1,82 @@
+function _pierceShadows(node, querySelectors, depth) {
+  if (undefined===depth) depth=0;
+
+
```
nit: rm extra blank here and below, a few more later in patch

shaper commented on dev/test/selenium-utils.js:
```diff
@@ -0,0 +1,82 @@
+ */
+
+
+let debug = true;
```
const

shaper commented on dev/test/specs/starter-test.js:
```diff
@@ -0,0 +1,192 @@
+describe('test basic arcs functionality', function() {
+  it('can use the restaurant demo flow', function() {
+    // TODO(smalls) need to spin up a server for this
+    let cdnBranch = 'http://localhost:8000/arcs-cdn/dev';
```
const

shaper commented on dev/test/selenium-utils.js:
```diff
@@ -0,0 +1,82 @@
+  while(l--){nextNodesArr[l]=nextNodes[l]};
+
+  let results = nextNodesArr.reduce((accumulator, currentValue) => {
+    !debug || console.log(`${Array(depth+1).join(' ')}descending to light ${currentValue} with selector ${remainingSelectors[0]}`);
```
nit: need a space before 'descending'

smalls committed:
prettier

shaper commented on dev/test/selenium-utils.js:
```diff
@@ -0,0 +1,82 @@
+
+  let results = nextNodesArr.reduce((accumulator, currentValue) => {
+    !debug || console.log(`${Array(depth+1).join(' ')}descending to light ${currentValue} with selector ${remainingSelectors[0]}`);
+    let lightNodes = _pierceShadows(currentValue, remainingSelectors, depth+1);
```
const

shaper commented on dev/test/selenium-utils.js:
```diff
@@ -0,0 +1,82 @@
+  var l=nextNodes.length, nextNodesArr = new Array(l);
+  while(l--){nextNodesArr[l]=nextNodes[l]};
+
+  let results = nextNodesArr.reduce((accumulator, currentValue) => {
```
const (or can just return the result of reduce directly)

shaper commented on dev/test/specs/starter-test.js:
```diff
@@ -0,0 +1,192 @@
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+let assert = require('assert');
```
We had to use the other form for our other browser tests, like:

`let assert = chai.assert`

Why is this ok here, maybe because this runs within Selenium rather than Mocha in a browser? Is this requiring the Chai assert lib or something else?

smalls replied:
Great question.

The require() call is specific to node. Because these are selenium tests, I don't think they can run in the browser, so this style is ok here.

I tried the other styles, both

```let assert = chai.assert```

and

```import chai from 'chai';```

One of those would be nice for consistency, but neither seems to work.

shaper commented on dev/test/specs/starter-test.js:
```diff
@@ -0,0 +1,192 @@
+ *   {id: <element-id>, text: <found text>}
+ */
+function searchElementsForText(elements, textQuery) {
+  let textToId = elements
```
const here (and for matches below), similar rest of file probs

shaper commented on dev/test/specs/starter-test.js:
```diff
@@ -0,0 +1,192 @@
+
+/** wait for the dancing dots to stop. */
+function waitForStillness() {
+  var element = pierceShadowsSingle(['arc-footer', 'x-toast[app-footer]', 'dancing-dots']);
```
Is const instead of var here and line 60 below ok?

shaper commented on dev/test/specs/starter-test.js:
```diff
@@ -0,0 +1,220 @@
+ *   {id: <element-id>, text: <found text>}
+ */
+function searchElementsForText(elements, textQuery) {
+  let textToId = elements.map(value => {
```
`const` here and below (similar rest of file probs, same for some `var` refs).

shaper commented on dev/test/wdio.conf.js:
```diff
@@ -0,0 +1,253 @@
+      // grid with only 5 firefox instances available you can make sure that not more than
+      // 5 instances get started at a time.
+      maxInstances: 5,
+      //
```
Something to write here?

shaper commented on dev/test/wdio.conf.js:
```diff
@@ -0,0 +1,253 @@
+      chromeOptions: {
+        args: [
+          // arcs note: comment this out to see the system running
+          '--headless'
```
For future consideration, would be handy if we could somehow do this also (see it rather than run headless) via a command line argument.

smalls replied:
Agreed, that'd be better. It's really nice for debugging to see it in action.

shaper reviewed (approved):
Just a few questions and noted some let/var that I think can be const. Some files changed while I was replying, it's cool.

smalls committed:
small cleanups from feedback
- Prefer const over let, let over var
- Use `if (debug)` rather than `!debug ||`; the former is clearer, and
  prettified it's similar in terms of space

smalls committed:
turn debugging off

smalls commented:
Part of https://github.com/PolymerLabs/arcs/issues/454
